torch/csrc/jit/testing/file_check.cpp

```diff
@@ -324,32 +314,63 @@ struct FileCheckImpl {
       throw std::runtime_error(ss.str());
     };
 
-    if (highlight_lineno >= source->num_lines()) {
-      construct_error_and_throw();
-    }
+    size_t search_start_offset = start_offset;
+    bool found_token_at_least_once = false;
+    size_t pos = search_start_offset;
+    while (pos < source->text().size()) {
+      auto pos = source->text().find(check.search_str_, search_start_offset);
+      if (pos == std::string::npos) {
+        break;
+      }
+
+      auto lineno = source->lineno_for_offset(pos);
+      auto col = pos - source->offset_for_line(lineno);
+      auto highlight_lineno = lineno + 1;
+
+      if (highlight_lineno >= source->num_lines()) {
+        construct_error_and_throw(pos);
+      }
 
-    auto highlight_start_offset =
-        source->offset_for_line(highlight_lineno) + col;
-    auto highlight_end_offset = std::min(
-        highlight_start_offset + check.search_str_.size(),
-        source->text().size());
+      auto highlight_start_offset =
+          source->offset_for_line(highlight_lineno) + col;
+      auto highlight_end_offset = std::min(
+          highlight_start_offset + check.search_str_.size(),
+          source->text().size());
 
-    if (highlight_end_offset >= source->text().size()) {
-      construct_error_and_throw();
+      if (highlight_end_offset >= source->text().size()) {
+        construct_error_and_throw(pos);
+      }
+
+      bool found_highlight = true;
+      for (auto pos = highlight_start_offset; pos < highlight_end_offset;
+           ++pos) {
+        if (source->text()[pos] != '~') {
+          found_highlight = false;
+        }
+      }
+
+      if (found_highlight) {
+        assertNotFind(
+            SourceRange(
+                source, highlight_start_offset - 1, highlight_start_offset),
+            "~",
+            check);
+        assertNotFind(
+            SourceRange(source, highlight_end_offset, highlight_end_offset + 1),
+            "~",
+            check);
+        return;
+      }
+
+      search_start_offset = pos + 1;
```

Comment:
Can we advance by the length of the search string here instead of 1? If execution reaches this point, then `pos` should point to the first character of an instance of the search string.

Reply:
Usually it is fine to advance by length of search string, but in case of the string being `aa` and the error message has a `aaaaaaaaaaaaa` we might miss matches.